llvm/lib/SYCLLowerIR/MutatePrintfAddrspace.cpp

```diff
@@ -0,0 +1,145 @@
+
+// Wrapper for the pass to make it working with the old pass manager
+class SYCLMutatePrintfAddrspaceLegacyPass : public ModulePass {
+public:
+  static char ID;
+  SYCLMutatePrintfAddrspaceLegacyPass() : ModulePass(ID) {
+    initializeSYCLMutatePrintfAddrspaceLegacyPassPass(
+        *PassRegistry::getPassRegistry());
+  }
+
+  // run the SYCLMutatePrintfAddrspace pass on the specified module
+  bool runOnModule(Module &M) override {
+    ModuleAnalysisManager MAM;
+    auto PA = Impl.run(M, MAM);
+    return !PA.areAllPreserved();
+  }
+
+private:
+  SYCLMutatePrintfAddrspacePass Impl;
+};
+
+Constant *getCASLiteral(IRBuilder<> &Builder, CallInst *CI) {
+  auto *Literal = cast<Constant>(CI->getArgOperand(0));
+  Type *LiteralType = Literal->getType();
+  // Generate/find a correct literal
+  auto *CASLiteralType = PointerType::get(LiteralType->getPointerElementType(),
+                                          ConstantAddrspaceID);
+  StringRef CASLiteralName(Literal->getName().str() + "._AS2");
+  return CI->getModule()->getOrInsertGlobal(
+      CASLiteralName, CASLiteralType, [&] {
+        StringRef LiteralValue;
+        getConstantStringInfo(Literal, LiteralValue);
+        GlobalVariable *GV = Builder.CreateGlobalString(
+            LiteralValue, CASLiteralName, ConstantAddrspaceID, CI->getModule());
+        GV->setLinkage(GlobalValue::LinkageTypes::InternalLinkage);
+        GV->setUnnamedAddr(GlobalValue::UnnamedAddr::None);
+        return GV;
+      });
+}
+
+FunctionCallee getCASPrintfFunction(Function &GenericASPrintfFunc,
+                                    Type *CASLiteralTy) {
+  auto *CASPrintfFuncTy = FunctionType::get(GenericASPrintfFunc.getReturnType(),
+                                            CASLiteralTy, /*isVarArg=*/true);
+  FunctionCallee CASPrintfFunc =
+      GenericASPrintfFunc.getParent()->getOrInsertFunction(
+          "_Z18__spirv_ocl_printfPU3AS2Kcz", CASPrintfFuncTy,
+          GenericASPrintfFunc.getAttributes());
+  auto *Callee = cast<Function>(CASPrintfFunc.getCallee());
+  Callee->setCallingConv(CallingConv::SPIR_FUNC);
+  Callee->setDSOLocal(true);
+  return CASPrintfFunc;
+}
+
+CallInst *buildCASPrintfCall(IRBuilder<> &Builder, FunctionCallee CASPrintfFunc,
+                             Constant *CASLiteral, CallInst *CI) {
+  SmallVector<Value *, 4> CallOperands(CI->data_operands_begin(),
+                                       CI->data_operands_end());
+  CallOperands[0] = CASLiteral;
+  Builder.SetInsertPoint(CI);
+  CallInst *CASCall = Builder.CreateCall(CASPrintfFunc, CallOperands,
+                                         CI->getName().str() + "._AS2");
+  CASCall->setTailCall(true);
+  return CASCall;
+}
+} // namespace
+
+PreservedAnalyses
+SYCLMutatePrintfAddrspacePass::run(Module &M, ModuleAnalysisManager &MAM) {
+  size_t ReplacedCallsCount = 0;
+  SmallVector<Function *, 1> FunctionsToDrop;
+
+  for (Function &F : M) {
+    if (!F.isDeclaration())
+      continue;
```

Comment:
GitHub UI doesn't allow me to add another comment without unresolving existing conversation on that line, so posting here.

We could use `.startswith("_Z18__spirv_ocl_printf")` instead of `.contains()` on the line below, as a one more optimization

Reply:
The mangling is different for the non-CAS version declared in the DPC++ headers. What's more important, I wouldn't really enjoy relying on the mangled name length (and therefore, on the header-level changes) within the pass.

Reply:
Number encoded there is a length of an unmangled function name, so it should be the same regardless of amount of arguments and their types